src/v/kafka/data/log_reader_config.h

```diff
@@ -0,0 +1,139 @@
+
+    /// \brief gurantees first_timestamp >= record_batch.first_timestamp
+    /// it is the std::lower_bound
+    std::optional<model::timestamp> first_timestamp;
+
+    /// abort source for read operations
+    model::opt_abort_source_t abort_source;
+
+    // used by log reader
+    size_t bytes_consumed{0};
+
+    // skipping consumer sets this bit if a read would cause max_bytes to be
+    // violated.  its used to signal to the log reader that even though consumed
+    // bytes hasn't reached max bytes that reading should still stop.
+    bool over_budget{false};
+
+    // do not let the reader go over budget even when that means that the reader
+    // will return no batches.
+    bool strict_max_bytes{false};
+
+    model::opt_client_address_t client_address;
+
+    // If set to true, when a gap is detected in the offset space, installs a
+    // `ghost_batch` to fill the gap with the same term as the next batch.
+    //
+    // The assumption here is that the gap was created by compaction, and that
+    // the corresponding ghost batches exactly fill the space, while preserving
+    // terms boundaries.
+    bool fill_gaps{false};
+
+    // If set to true, the offsets returned will be translated from Redpanda
+    // offset to its data offset, as dictated by the underlying log's offset
+    // translator types.
+    //
+    // NOTE: the translation refers only to the returned batches, not to the
+    // input min/max offset bounds. Callers are expected to account for inputs
+    // separately.
+    translate_offsets translate_offsets{false};
+
+    log_reader_config(
+      model::offset start_offset,
+      model::offset max_offset,
+      size_t min_bytes,
+      size_t max_bytes,
+      std::optional<model::record_batch_type> type_filter,
+      std::optional<model::timestamp> time,
+      model::opt_abort_source_t as,
+      model::opt_client_address_t client_addr = std::nullopt)
+      : start_offset(start_offset)
+      , max_offset(max_offset)
+      , min_bytes(min_bytes)
+      , max_bytes(max_bytes)
+      , type_filter(type_filter)
+      , first_timestamp(time)
+      , abort_source(as)
+      , client_address(std::move(client_addr)) {}
+
+    /**
+     * Read offsets [start, end].
+     */
+    log_reader_config(
+      model::offset start_offset,
+      model::offset max_offset,
+      model::opt_abort_source_t as = std::nullopt,
+      model::opt_client_address_t client_addr = std::nullopt)
+      : log_reader_config(
+          start_offset,
+          max_offset,
+          0,
+          std::numeric_limits<size_t>::max(),
+          std::nullopt,
+          std::nullopt,
+          as,
+          std::move(client_addr)) {}
+
+    friend std::ostream& operator<<(std::ostream& o, const log_reader_config&);
+};
+
+// Empty, invalid reader config which is sometimes useful as a placeholder
+// since log_reader_config doesn't have a default constructor.
+static const log_reader_config empty_reader_config{{}, {}};
```

Comment:
Didn't yet read through the use-case of this but I believe this should be marked inline